src/shogun/machine/GLM.h

```diff
@@ -0,0 +1,296 @@
+/*
+ * This software is distributed under BSD 3-clause license (see LICENSE file).
+ *
+ * Both the documentation and the code is heavily inspired by pyGLMnet:
+ * https://github.com/glm-tools/pyglmnet/
+ *
+ * Author: Tej Sukhatme
+ */
+
+#include <shogun/lib/config.h>
+
+#include <shogun/distributions/Distribution.h>
+#include <shogun/lib/common.h>
+#include <shogun/machine/IterativeMachine.h>
+#include <shogun/machine/LinearMachine.h>
+#include <shogun/machine/Machine.h>
+#include <shogun/mathematics/RandomMixin.h>
+#include <shogun/mathematics/linalg/LinalgNamespace.h>
+#include <shogun/optimization/ConstLearningRate.h>
+#include <shogun/optimization/ElasticNetPenalty.h>
+#include <shogun/optimization/FirstOrderCostFunction.h>
+#include <shogun/optimization/GradientDescendUpdater.h>
+
+namespace shogun
+{
+	enum GLM_DISTRIBUTION
+	{
+		POISSON
+	};
+
+	class DotFeatures;
+	class Features;
+	class RegressionLabels;
+	class GLMCostFunction;
+
+	/** @brief Class for estimating regularized generalized linear models (GLM).
+	 *   The regularized GLM minimizes the penalized negative log likelihood:
+	 *
+	 *  This uses Elastic-net penalty which defaults to the ridge penalty when
+	 *  alpha = 0 and defaults to the lasso penalty when alpha = 1.
+	 *
+	 * */
+	class GLM : public RandomMixin<IterativeMachine<LinearMachine>>
+	{
+	public:
+		friend class GLMCostFunction;
+
+		/** default constructor */
+		GLM();
+
+		GLM(GLM_DISTRIBUTION distribution, float64_t alpha, float64_t lambda,
+		    float64_t learning_rate, int32_t max_iterations,
+		    float64_t tolerance, float64_t eta);
+
+		/** destructor */
+		~GLM() override
+		{
+		}
+
+		/** apply linear machine to data
+		 * for regression problem
+		 *
+		 * @param data (test)data to be classified
+		 * @return classified labels
+		 */
+		std::shared_ptr<RegressionLabels>
+		apply_regression(std::shared_ptr<Features> data = NULL) override;
+
+		/** Returns the name of the SGSerializable instance.  It MUST BE
+		 *  the CLASS NAME without the prefixed `C'.
+		 *
+		 * @return name of the SGSerializable
+		 */
+		const char* get_name() const override
+		{
+			return "GLM";
+		}
+
+		static char* glm_enum_to_string(GLM_DISTRIBUTION distribution)
+		{
+			switch (distribution)
+			{
+			case POISSON:
+				return "Poisson";
+			default:
+				return "Invalid distribution";
+			}
+		}
+
+		virtual EProblemType get_machine_problem_type() const override
+		{
+			return PT_REGRESSION;
+		}
+
+		protected:
+		void init_model(const std::shared_ptr<Features> data);
+
+		void iteration() override;
+
+	private:
+		/** Distribution type */
+		GLM_DISTRIBUTION distribution = POISSON;
```

Comment:
The serialization issue is caused by this `enum` here. Apparently, RapidJSON does not like it. I don't even know if this is a bug of the serialization library or something else.

Reply:
in what way? Is it not serialising it properly? or is not deserialising it?

Reply:
It triggers a RapidJSON's assertion and the program crashes. Therefore, nothing is serialized.

Reply:
that is odd, because there are other classes with `enum`s that do not have issues.. Could you show the assertion? Is this in debug mode?

Reply:
```native_example: /usr/include/rapidjson/writer.h:448: void rapidjson::Writer<OutputStream, SourceEncoding, TargetEncoding, StackAllocator, writeFlags>::Prefix(rapidjson::Type) [with OutputStream = OutputStreamAdapter; SourceEncoding = rapidjson::UTF8<>; TargetEncoding = rapidjson::UTF8<>; StackAllocator = rapidjson::CrtAllocator; unsigned int writeFlags = 2]: Assertion `type == kStringType' failed.```

The assertion is this one.

Reply:
see comment about `SG_ADD(&distribution, ...`